### PR TITLE
Header updates for flint 3

### LINKED ISSUE #2972: Build fails with Flint 3.0.0
I tried building with the new Flint release and ran into an error:
```cpp
/home/mahrud/Projects/M2/quickfix/M2/Macaulay2/e/aring-zz-flint.hpp:46:38: fatal error: use of undeclared identifier 'fmpz_get_ui'; did you mean '__gmpz_get_ui'?
    return static_cast<unsigned int>(fmpz_get_ui(&a));
                                     ^~~~~~~~~~~
                                     __gmpz_get_ui
/home/linuxbrew/.linuxbrew/opt/gmp/include/gmp.h:1774:1: note: '__gmpz_get_ui' declared here
mpz_get_ui (mpz_srcptr __gmp_z) __GMP_NOTHROW
^
/home/linuxbrew/.linuxbrew/opt/gmp/include/gmp.h:835:20: note: expanded from macro 'mpz_get_ui'
#define mpz_get_ui __gmpz_get_ui
                   ^
1 error generated.
```
I didn't spend any time debugging, but this should be treated soon because once https://github.com/Homebrew/homebrew-core/pull/151985 is merged M2 will stop working if you upgrade Flint via Homebrew, for instance.

### PR DESCRIPTION
**Pull request by mahrud**
- updated flint submodule to v3.0.0
- bumped minimum flint version to 2.8.5
- corrected inclusion of flint headers for v3.0.0
- switched to FLINT_VERSION instead of flint_version
- renamed flint2 submodule to flint
- bumped factory to 4.3.0 and patched ntl and factory

Closes #2972

### CODE REVIEW

**Comment by mahrud** [comment]
The autotools build doesn't seem to use the submodule and instead downloads a very old version: 2.8.4! Someone else should perhaps fix that.
```
make[2]: Entering directory '/home/runner/work/M2/M2/M2/BUILD/build/libraries/flint'
(cd /home/runner/work/M2/M2/M2/BUILD/build/../../BUILD/tarfiles && wget http://macaulay2.com/Downloads/OtherSourceCode/flint-2.8.4.tar.gz  && case "http://macaulay2.com/Downloads/OtherSourceCode" in http:*) if file /home/runner/work/M2/M2/M2/BUILD/build/../../BUILD/tarfiles/flint-2.8.4.tar.gz | grep HTML ; then echo tried to fetch file, but HTML file returned instead >&2 ; exit 1 ; fi ;; esac)
--2023-10-24 07:38:45--  http://macaulay2.com/Downloads/OtherSourceCode/flint-2.8.4.tar.gz
Resolving macaulay2.com (macaulay2.com)... 172.67.222.115, 104.21.62.93, [26](https://github.com/Macaulay2/M2/actions/runs/6623766282/job/17991405631#step:11:27)06:4700:130:436c:6f75:6466:6c61:7265
Connecting to macaulay2.com (macaulay2.com)|172.67.222.115|:80... connected.
HTTP request sent, awaiting response... 301 Moved Permanently
Location: https://macaulay2.com/Downloads/OtherSourceCode/flint-2.8.4.tar.gz [following]
--2023-10-24 07:38:45--  https://macaulay2.com/Downloads/OtherSourceCode/flint-2.8.4.tar.gz
Connecting to macaulay2.com (macaulay2.com)|172.67.222.115|:443... connected.
HTTP request sent, awaiting response... 200 OK
Length: 521[28](https://github.com/Macaulay2/M2/actions/runs/6623766282/job/17991405631#step:11:29)88 (5.0M) [application/x-gzip]
Saving to: ‘flint-2.8.4.tar.gz’
```

**Comment by mahrud** [comment]
Thanks! I believe you should have push access to my branch.

There seems to be a failure involving `1/i` in `k = toField(QQ[i]/(i^2+1));`. I had to patch ntl and factory so I could build them with the new flint.

I vaguely recall there was a regression due to factory 4.3.0? Was that ever resolved?

**Comment by mahrud** [comment]
Still can't reproduce this on Fedora. @d-torrance could you give it a try on Ubuntu? This is the error on GitHub builds:
```m2
i9 : k = toField (QQ[i]/(i^2+1))

o9 = k

o9 : PolynomialRing

i10 : 1/i
stdio:10:2:(3): error: either element not invertible, or no method available to compute its inverse
```

**Comment by mahrud** [comment]
Looks like the block that had delayed homebrew bumping to flint 3 is resolved.

**Comment by mahrud** [comment]
> I'll volunteer to be "someone". :)

Have you had a chance to look into this?

**Comment by mikestillman** [comment]
@mahrud

**Comment by mikestillman** [comment]
Not sure how I closed this pull request.  Some hotkey sequence!?  Anyway, it is reopened.  But now it is re-doing the tests.

**Comment by d-torrance** [comment]
> Have you had a chance to look into this?

I've started, but it's not working yet.

**Comment by mikestillman** [comment]
@d-torrance Does this pull request include the changes needed for the autotools build to work with flint version 3? 

@mahrud Actually, without this PR, I can no longer build Macaulay2.

**Comment by d-torrance** [comment]
> Does this pull request include the changes needed for the autotools build to work with flint version 3?

*Work* with flint 3, probably.  Mahrud's updates to things like `#include`'s should do the trick.  I haven't tried building against a system installation of flint 3 yet.

*Build* flint 3, not yet.  I have a branch I'm working on where it successfully builds flint 3 from the submodule, compiles, but fails when linking.  So it needs more work.  :)

**Comment by mikestillman** [comment]
@d-torrance, The autotools builds on github test seem to use flint 2.8.4, and have the following error:
```
../../../../Macaulay2/e/aring-zzp-flint.hpp:21:10: fatal error: flint/nmod.h: No such file or directory
   21 | #include <flint/nmod.h>   // for nmod_neg, nmod_add, nmod_div, nmod_mul
      |          ^~~~~~~~~~~~~~
```

Strangely, it works, (or seems to work) on my Apple M1 laptop (probably with flint 3).  So, do we need to change the autotools build to grab flint version 3?

**Comment by mikestillman** [comment]
@d-torrance @DanGrayson @mahrud How do I find the artifacts?  I want to see what the error is in BeginningMacaulay2 install, I thought I used to be able to see these artifacts, but now I don't see how!?

**Comment by d-torrance** [comment]
> @d-torrance, The autotools builds on github test seem to use flint 2.8.4, and have the following error:
> 
> ```
> ../../../../Macaulay2/e/aring-zzp-flint.hpp:21:10: fatal error: flint/nmod.h: No such file or directory
>    21 | #include <flint/nmod.h>   // for nmod_neg, nmod_add, nmod_div, nmod_mul
>       |          ^~~~~~~~~~~~~~
> ```
> 
> Strangely, it works, (or seems to work) on my Apple M1 laptop (probably with flint 3). So, do we need to change the autotools build to grab flint version 3?

`nmod.h` must be new -- it's not present in 2.8.4.

I think the autotools build should grab flint 3 if it's *building* flint, but I think we should also support older system installations of flint as well.  These pre-3.0.0 versions are going to be floating around in Ubuntu LTS's for years, for example.  (Plus flint 3 hasn't been packaged for Debian yet, and it's just a few months before the freeze for the next Ubuntu LTS.)

**Comment by mikestillman** [comment]
@d-torrance @mahrud Doug, from what you are saying, it appears that we might need to ifdef the header files for flint, and maybe some functions too...

**Comment by d-torrance** [comment]
> * bumped minimum flint version to 2.8.5

What's special about 2.8.5?

**Comment by mikestillman** [comment]
@d-torrance @mahrud One issue is that I need to accept this soon, as it stalls my working on code, since my builds now fail after I upgraded brew tap.  But it looks like it is just header files, and not changed function names.

I guess I can downgrade flint to 2.8.4 (this version seems to work with autotools at least).  Any suggestions how we should fix this PR?

I also still don't know why the one error is failing either.  I expect the error is the one @mahrud mentions above, but I'm not sure why that would be failing.

**Comment by mikestillman** [comment]
@d-torrance What versions of flint should we support?

**Comment by d-torrance** [comment]
> What versions of flint should we support?

I'd prefer to stick with >= 2.6.3 if possible.

**Comment by mikestillman** [comment]
> > What versions of flint should we support?
> 
> I'd prefer to stick with >= 2.6.3 if possible.

@d-torrance @mahrud Do you know which version the header files changed?  Is it 2.8.5, or 3.0? (or something else)

**Comment by d-torrance** [comment]
>  Do you know which version the header files changed? Is it 2.8.5, or 3.0? (or something else)

It was a couple years ago, so maybe it was 2.8.5:

```
commit d25e372ecdc7674bcb90c11a757a6bf67b30b80f
Author: fredrik <fredrik.johansson@gmail.com>
Date:   Wed Nov 24 12:32:31 2021 +0100

    split nmod_vec.h -> nmod.h, nmod_vec.h and document a couple of things
```

**Comment by mahrud** [comment]
> I'm assuming there's a cmake equivalent of `AC_CHECK_HEADERS` that would define `HAVE_FLINT_NMOD_H` if it's present?

Sure, you can add something like `check_include_files(flint/nmod.h HAVE_FLINT_NMOD_H)` in a new block like this:
https://github.com/Macaulay2/M2/blob/86d6c70f8746e6b13d95c53ad20d880e0d60a8dc/M2/cmake/check-libraries.cmake#L346-L353
Be sure to also add a block like:
https://github.com/Macaulay2/M2/blob/86d6c70f8746e6b13d95c53ad20d880e0d60a8dc/M2/include/M2/config.h.cmake#L54-L55

**Comment by mahrud** [comment]
> How do I find the artifacts? I want to see what the error is in BeginningMacaulay2 install, I thought I used to be able to see these artifacts, but now I don't see how!?

There's a button on the top left when you're looking at the action logs.
image

The error is this [one](https://github.com/Macaulay2/M2/pull/2973#issuecomment-1778504371).

**Comment by d-torrance** [comment]
I just pushed a few commits adding a `#define` that tells us whether `flint/nmod.h` exists.

File: M2/cmake/check-libraries.cmake [hunk 1]
```diff
@@ -352,6 +352,13 @@ else()
   unset(FACTORY_STREAMIO CACHE)
 endif()
 
+if(FLINT_FOUND)
+  check_include_files(flint/nmod.h HAVE_FLINT_NMOD_H)
+else()
+  # if flint isn't found, then we'll build a new version that has nmod.h
+  set(HAVE_FLINT_NMOD_H 1)
```
Comment (mahrud) [inline M2/cmake/check-libraries.cmake:359]: Don't do this, it'll mess things up if the user wants to switch from building with Flint 3 to another version or vice versa. Just use `unset(HAVE_FLINT_NMOD_H CACHE)`, and if Flint is being built, then the variable will be set after that's done.

Comment (d-torrance) [reply M2/cmake/check-libraries.cmake]: Ok, cool.  I wasn't sure if what I was doing was the best approach.  :)  I just force-pushed a new version of this commit.

**Comment by mahrud** [comment]
It's strange that this error is kind of sporadic:
```cpp
../../../../Macaulay2/d/python-c.c:108:38: fatal error: no member named 'ob_digit' in 'struct _longobject'
    mpz_set_si(z, -(sdigit)templong->ob_digit[0]);
                           ~~~~~~~~  ^
1 error generated.
```

**Comment by d-torrance** [comment]
> It's strange that this error is kind of sporadic:
> 
> ```c++
> ../../../../Macaulay2/d/python-c.c:108:38: fatal error: no member named 'ob_digit' in 'struct _longobject'
>     mpz_set_si(z, -(sdigit)templong->ob_digit[0]);
>                            ~~~~~~~~  ^
> 1 error generated.
> ```

It's a Python 3.12 thing.  I've fixed it in a branch so that it compiles with Python 3.12, but now I'm getting a weird segfault when exiting Macaulay2 and finalizing Python.  Hopefully I'll have a working PR soon!

File: M2/include/M2/config.h.cmake [hunk 2]
```diff
@@ -54,6 +54,9 @@
 /* whether factory was built with --enable-streamio */
 #cmakedefine FACTORY_STREAMIO 1
 
+/* whether we have the flint header file nmod.h */
+#cmakedefine HAVE_FLINT_NMOD_H 1
```
Comment (mahrud) [inline M2/include/M2/config.h.cmake:58]: This is not being defined correctly on GitHub:
```cpp
/* whether we have the flint header file nmod.h */
/* #undef HAVE_FLINT_NMOD_H */
```
Not sure why.

File: M2/cmake/check-libraries.cmake [hunk 3]
```diff
@@ -352,6 +352,12 @@ else()
   unset(FACTORY_STREAMIO CACHE)
 endif()
 
+if(FLINT_FOUND)
+  check_include_files(flint/nmod.h HAVE_FLINT_NMOD_H)
```
Comment (mahrud) [inline M2/cmake/check-libraries.cmake:356]: Ohh, we need something like this:
```cmake
  set(CMAKE_REQUIRED_INCLUDES "${FLINT_INCLUDE_DIR}")
```

Comment (mahrud) [reply M2/cmake/check-libraries.cmake:356]: Otherwise it's just checking system directories.

Comment (d-torrance) [reply M2/cmake/check-libraries.cmake:356]: Gotcha -- I was wondering if we needed that or not!

**Comment by mahrud** [comment]
The cmake error is totally bizarre, and I can't reproduce it any way on my system, but also it's probably not worth wasting more time on it if a less than ideal workaround fixes it.

**Comment by mikestillman** [comment]
@mahrud @d-torrance Thanks for your work on this!

I see two things failing here at the moment: in engine.cpp, nmod_neg cannot be found (ubuntu), and on the mac builds (both autotools and cmake), it looks like python crashes at the end of the program.  As soon as we (probably the both of you...) can fix those, I would like to get this accepted soon.  For the python problem, I suppose these builds are including python support?  Is the other one just a missing header file?

**Comment by d-torrance** [comment]
> it looks like python crashes at the end of the program. ... For the python problem, I suppose these builds are including python support?

It's more than just the call to `Py_FinalizeEx` and the end ...  loading the `Python` package and doing pretty much anything other than  dealing with booleans or small integers (so I think anything where Python actually has to allocate some memory) results in a segfault using Python 3.12.  Disabling Python support on the macOS GitHub builds would be a quick bandaid fix for now until I can figure out what's going on.

**Comment by mahrud** [comment]
There's more examples failing, now `getIdealFromNoetherianOperators`:
```m2
i1 : R = QQ[x_1,x_2,x_3,x_4]

o1 = R

o1 : PolynomialRing

i2 : MM = matrix {{x_3,x_1,x_2},{x_1,x_2,x_4}}

o2 = | x_3 x_1 x_2 |
     | x_1 x_2 x_4 |

             2       3
o2 : Matrix R  <--- R

i3 : P = minors(2,MM)

               2                           2
o3 = ideal (- x  + x x , - x x  + x x , - x  + x x )
               1    2 3     1 2    3 4     2    1 4

o3 : Ideal of R

i4 : M = ideal{x_1^2,x_2^2,x_3^2,x_4^2}

             2   2   2   2
o4 = ideal (x , x , x , x )
             1   2   3   4

o4 : Ideal of R

i5 : Q = joinIdeals(P,M);

o5 : Ideal of R

i6 : L1 = noetherianOperators(Q) -- A set of Noetherian operators
-- SIGSEGV
-* stack trace, pid: 72281
 0# stack_trace(std::ostream&, bool) at /home/runner/work/M2/M2/M2/Macaulay2/bin/main.cpp:123
 1# segv_handler at /home/runner/work/M2/M2/M2/Macaulay2/bin/main.cpp:240
 2# 0x00007F7B77842520 in /lib/x86_64-linux-gnu/libc.so.6
 3# ListItem<Factor<CanonicalForm> >::getItem() in /lib/x86_64-linux-gnu/libsingular-factory-4.2.1.so
 4# List<Factor<CanonicalForm> >::getFirst() const in /lib/x86_64-linux-gnu/libsingular-factory-4.2.1.so
 5# 0x00007F7B791BB700 in /lib/x86_64-linux-gnu/libsingular-factory-4.2.1.so
 6# irrCharSeries(List<CanonicalForm> const&) in /lib/x86_64-linux-gnu/libsingular-factory-4.2.1.so
 7# rawCharSeries at /home/runner/work/M2/M2/M2/Macaulay2/e/interface/factory.cpp:843
 8# interface_rawCharSeries at /home/runner/work/M2/M2/M2/Macaulay2/d/interface.dd:1692
```

**Comment by mahrud** [comment]
Compiling is finishing everywhere now, but some examples and tests seem to be failing.

**Comment by mahrud** [comment]
Since M2 at least compiles with flint 3 now, I reduced minimum required version of flint back to 2.8.3 so that tests pass here and this can be merged. That said, more work remains to:

1. re-enable Python on GitHub
2. fix possible bugs with examples and tests using flint 3

**Comment by mikestillman** [comment]
@d-torrance @mahrud Thanks for all the work you have done to fix this issue!

@DanGrayson I am going to accept this pull request (with rebase and merge).

**Comment by d-torrance** [comment]
I'm getting https://github.com/Macaulay2/M2/issues/1674 again on a bunch of PPA builds after this was merged.  Hopefully it's an easy fix like https://github.com/Macaulay2/M2/pull/1677 again.

**Pull request merged**